Add get_blob_stream (prefer GetRawBlob)

0verflowme opened the pull request:
This adds a higher-level blob streaming API that prefers the newer GetRawBlob method (when supported by the negotiated proxy protocol version) and otherwise falls back to GetBlob.

Motivation: https://github.com/bootc-dev/bootc/issues/1424

0verflowme committed:
Add get_blob_stream (prefer GetRawBlob)

Signed-off-by: Priyanshu Kumar <priyanshu.kumar@broadcom.com>

cgwalters reviewed (commented):
Thanks for working on this! Just as a first pass I think what would help build confidence here is ensuring in CI we're always testing both the old and new code.

One suggestion: A CI run that tests in `quay.io/almalinuxorg/almalinux-bootc:10.0` as that uses `skopeo version 1.18.1` which predates https://github.com/containers/skopeo/releases/tag/v1.19.0 when this feature appeared.

0verflowme committed:
ci: test get_blob_stream against old/new skopeo

Run tests in containers with skopeo 1.18 and >=1.19 to exercise both GetBlob fallback and GetRawBlob paths.

Signed-off-by: Priyanshu Kumar <priyanshu.kumar@broadcom.com>

cgwalters commented on src/imageproxy.rs:
```diff
@@ -88,6 +92,187 @@ impl From<rustix::io::Errno> for Error {
+            v @ std::task::Poll::Ready(Ok(_)) => {
+                let after = buf.filled().len();
+                debug_assert!(after >= before);
+                let delta = after - before;
```
How about `let delta = after.checked_sub(before).unwrap()`

cgwalters commented on src/imageproxy.rs:
```diff
@@ -739,6 +937,69 @@ impl ImageProxy {
+            return fallback_to_get_blob().await;
+        }
+
+        match self.get_raw_blob(img, digest).await {
```
We can just `?` here

cgwalters commented on src/imageproxy.rs:
```diff
@@ -88,6 +92,187 @@ impl From<rustix::io::Errno> for Error {
+pub struct BlobStream<'a> {
+    pub source: BlobStreamSource,
+    pub expected_size: u64,
+    pub reported_size: Option<u64>,
```
We *always* have a known size in OCI images.

cgwalters commented on src/imageproxy.rs:
```diff
@@ -88,6 +92,187 @@ impl From<rustix::io::Errno> for Error {
+
+#[derive(Debug)]
+enum VerifiedBlobReadResult {
+    Complete { nbytes: u64, digest_hex: String },
```
Let's add type safety to digest_hex using https://docs.rs/oci-spec/latest/oci_spec/image/struct.Digest.html

cgwalters commented on src/imageproxy.rs:
```diff
@@ -88,6 +92,187 @@ impl From<rustix::io::Errno> for Error {
+                debug_assert!(after >= before);
+                let delta = after - before;
+                if delta > 0 {
+                    let chunk = &buf.filled()[before..after];
```
I am a bit confused by this...it seems like we're trying to avoid skipping the start of the buffer, but why do we need to do that? Can't we just consume `filled()` each time?

Also isn't this missing calling `clear()`?

(I am not super familiar with implementing AsyncRead this way to be clear)

0verflowme replied:
`ReadBuf::filled()` is the entire filled region, not “the bytes from this poll”. An `AsyncRead` impl is expected to append to the caller-provided `ReadBuf`, so it may already contain filled bytes when this wrapper is polled. I added a short code comment for this

cgwalters reviewed (commented):
Thanks so much for working on this! A second pass review

0verflowme committed:
Address blob stream review feedback

Assisted-by: Codex (GPT-5)
Signed-off-by: Priyanshu Kumar <priyanshu.kumar@broadcom.com>
Signed-off-by: Colin Walters <walters@verbum.org>

0verflowme commented on src/imageproxy.rs:
```diff
@@ -88,6 +92,187 @@ impl From<rustix::io::Errno> for Error {
+    inner: R,
+    nbytes: u64,
+    hasher: Option<Hasher>,
+    completion: Option<oneshot::Sender<VerifiedBlobReadResult>>,
```
Added a comment explaining the control flow.

cgwalters reviewed (dismissed):
Thanks!
